### PR TITLE
[subset] During os2 subsetting update ulUnicodeRange[]c

### PR DESCRIPTION
**Pull request by garretrieger**

### CODE REVIEW

File: src/hb-ot-os2-unicode-ranges.hh [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+#ifndef HB_OT_OS2_UNICODE_RANGES_HH
+#define HB_OT_OS2_UNICODE_RANGES_HH
+
+namespace OT {
+
+struct Range
+{
+  unsigned int start;
+  unsigned int end;
```
Comment (behdad) [inline src/hb-ot-os2-unicode-ranges.hh]: hb_codepoint_t please for start and end.

File: src/hb-ot-os2-table.hh [hunk 2]
```diff
@@ -67,11 +67,39 @@ struct os2
...
+  inline void _update_unicode_ranges (const hb_prealloced_array_t<hb_codepoint_t> &codepoints,
+                                      HBUINT32 ulUnicodeRange[4]) const
+  {
+    for (unsigned int i = 0; i < 4; i++)
+      ulUnicodeRange[i].set (0);
+
+    for (unsigned int i = 0; i < codepoints.len; i++)
+    {
+      hb_codepoint_t cp = codepoints[i];
+      int bit = hb_get_unicode_range_bit (cp);
```
Comment (behdad) [inline src/hb-ot-os2-table.hh]: unsigned int

File: src/hb-ot-os2-table.hh [hunk 3]
```diff
@@ -67,11 +67,39 @@ struct os2
...
+      if (bit >= 0 && bit < 128) {
```
Comment (behdad) [inline src/hb-ot-os2-table.hh]: Opening bracket on new lnie.

File: src/hb-ot-os2-unicode-ranges.hh [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+  { 0xE0100,  0xE01EF,  91}, // Variation Selectors Supplement
+  { 0xF0000,  0xFFFFD,  90}, // Private Use (plane 15)
+  {0x100000, 0x10FFFD,  90}, // Private Use (plane 16)
+};
+
+static int
+_compare_range (const void *_key, const void *_item, void *_arg)
+{
+  hb_codepoint_t *cp = (hb_codepoint_t *) _key;
+  Range *range = (Range *) _item;
```
Comment (behdad) [inline src/hb-ot-os2-unicode-ranges.hh]: const Range *

File: src/hb-ot-os2-unicode-ranges.hh [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+  { 0xE0000,  0xE007F,  92}, // Tags
+  { 0xE0100,  0xE01EF,  91}, // Variation Selectors Supplement
+  { 0xF0000,  0xFFFFD,  90}, // Private Use (plane 15)
+  {0x100000, 0x10FFFD,  90}, // Private Use (plane 16)
+};
+
+static int
+_compare_range (const void *_key, const void *_item, void *_arg)
+{
+  hb_codepoint_t *cp = (hb_codepoint_t *) _key;
```
Comment (behdad) [inline src/hb-ot-os2-unicode-ranges.hh]: cosnt

Comment (behdad) [reply src/hb-ot-os2-unicode-ranges.hh]: Better yet. Dereference the pointer here and carry codepoint around.

File: src/hb-ot-os2-unicode-ranges.hh [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+namespace OT {
+
+struct Range {
+  unsigned int start;
+  unsigned int end;
+  unsigned int bit;
+};
+
+/* Note: The contents of this array was generated using util/generate-unicode-ranges.py. */
+static Range os2UnicodeRangesSorted[] = {
```
Comment (behdad) [inline src/hb-ot-os2-unicode-ranges.hh]: Ideally this should be moved to a .cc file.

File: util/generate-unicode-ranges.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+
+# Generates the code for a sorted unicode range array as used in hb-ot-os2-unicode-ranges.hh
+# Input is a tab seperated list of unicode ranges from the otspec.
```
Comment (behdad) [inline util/generate-unicode-ranges.py]: Link to spec please.

Comment (behdad) [reply util/generate-unicode-ranges.py]: util/ is not for these things. Should go in src/ and named 'gen-....py'

File: util/generate-unicode-ranges.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+import io
+import re
+import sys
+
+reload(sys)
+sys.setdefaultencoding('utf-8')
+
+print (u"""static Range os2UnicodeRangesSorted[] = {""")
```
Comment (behdad) [inline util/generate-unicode-ranges.py]: "{" on new line.

File: src/hb-ot-os2-unicode-ranges.hh (hunk 1 above)
Comment (garretrieger) [reply src/hb-ot-os2-unicode-ranges.hh]: done

File: src/hb-ot-os2-table.hh (hunk 2 above)
Comment (garretrieger) [reply src/hb-ot-os2-table.hh]: done

File: src/hb-ot-os2-table.hh (hunk 3 above)
Comment (garretrieger) [reply src/hb-ot-os2-table.hh]: done

File: src/hb-ot-os2-unicode-ranges.hh (hunk 4 above)
Comment (garretrieger) [reply src/hb-ot-os2-unicode-ranges.hh]: done

File: src/hb-ot-os2-unicode-ranges.hh (hunk 5 above)
Comment (garretrieger) [reply src/hb-ot-os2-unicode-ranges.hh]: done

File: src/hb-ot-os2-unicode-ranges.hh (hunk 6 above)
Comment (garretrieger) [reply src/hb-ot-os2-unicode-ranges.hh]: Unfortunately doesn't work well with the boundary between libharfbuzz and libharfbuzz-subset. As discussed offline will leave here for now.

File: util/generate-unicode-ranges.py (hunk 7 above)
Comment (garretrieger) [reply util/generate-unicode-ranges.py]: done

File: util/generate-unicode-ranges.py (hunk 8 above)
Comment (garretrieger) [reply util/generate-unicode-ranges.py]: done

**Comment by behdad** [comment]
lgtm; please merge when bots agree.

**Pull request merged**